### PR TITLE
Exiled Component System Overhaul, DynamicEvents & StateMachine APIs, `Exiled::Updater` rework.

### PR DESCRIPTION
**Pull request by NaoUnderscore**
Added `DynamicEvents` API.
Added `EBehaviour` component.
Added `StateMachine` API
Reworked `Exiled::Updater`.

Going to explain those additions/changes later today.

### CODE REVIEW

File: Exiled.Loader/Updater.cs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+                };
+                installerProcess.BeginErrorReadLine();
+
+                installerProcess.WaitForExit();
+
+                Log.Info($"Installer exit code: {installerProcess.ExitCode}");
+
+                if (installerProcess.ExitCode == 0)
+                {
+                    Log.Info("Auto-update complete, round will be restarted the next round!");
```
Comment (VALERA771) [inline Exiled.Loader/Updater.cs:239]:
```suggestion
                    Log.Info("Auto-update complete, server will be restarted the next round!");
```

**Review by VALERA771 — APPROVED** [review]
Lgtm

**Comment by louis1706** [comment]
image
we need to change the build check

**Comment by NaoUnderscore** [comment]
> image we need to change the build check

Already fixed, this PR must be merged before.

**Comment by NaoUnderscore** [comment]
> Lgtm. You should also edit exactly Exiled/Exiled.nuspec. There is still updater

Where?

File: Exiled.Loader/Models/ExiledLib.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        /// A value that indicates the relative order of the objects being compared.
+        /// The return value has these meanings: Value Meaning Less than zero This instance precedes other in the sort order.
+        /// Zero This instance occurs in the same position in the sort order as other.
+        /// Greater than zero This instance follows other in the sort order.
+        /// </returns>
+        public int CompareTo(object obj)
+        {
+            if (obj is ExiledLib l)
+                return CompareTo(l);
+            else
```
Comment (louis1706) [inline Exiled.Loader/Models/ExiledLib.cs:56]: useless else

File: Exiled.Loader/Models/ExiledLib.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                return 0;
+        }
```
Comment (louis1706) [inline Exiled.Loader/Models/ExiledLib.cs:58]:
```suggestion
        public int CompareTo(object obj) => obj is ExiledLib l ? CompareTo(l) : 0
```

**Review by louis1706 — COMMENTED** [review]
~~i will suggest an other change ~~

**Review by louis1706 — APPROVED** [review]

**Review by louis1706 — APPROVED** [review]

File: Exiled.API/Features/Core/EBehaviour.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            BehaviourUpdate_Implementation();
+        }
+
+        /// <inheritdoc/>
+        protected override void OnEndPlay()
+        {
+            base.OnEndPlay();
+
+            if (Owner is null)
+                return;
```
[the remark this answers is not in the document]
Comment (VALERA771) [reply Exiled.API/Features/Core/EBehaviour.cs:50]: Maybe here sould be `Destroy` method as in other?

[the remark this answers is not in the document]
Comment (NaoUnderscore) [reply Exiled.API/Features/Core/EBehaviour.cs:50]: no

**Review by louis1706 — APPROVED** [review]

**Pull request merged**